tfx/experimental/pipeline_testing/dummy_component_launcher.py

```diff
@@ -0,0 +1,102 @@
+# Lint as: python2, python3
+# Copyright 2019 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""dummy component launcher which launches python executors and dummy executors in process."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import abc
+from typing import Any, Dict, List, Text, Type
+import absl
+
+from tfx import types
+from tfx.components.base import base_executor
+from tfx.orchestration.launcher.in_process_component_launcher import InProcessComponentLauncher
+from tfx.experimental.pipeline_testing.dummy_executor import BaseDummyExecutor
+
+class BaseDummyComponentLauncher(InProcessComponentLauncher):
+  """Responsible for launching a dummy executor.
+
+  The executor will be launched in the same process of the rest of the
+  component, i.e. its driver and publisher.
+  """
+  def __init__(self, **kwargs):
```

Comment:
Base..Launcher __init__(..) should call set_dummy_executors(), because we don't want users to define their own __init__ function.

Reply:
I also tried 
```

def create_dummy_launcher_class(record_dir, component_ids, custom_component_map):

  class DummyComponentLauncher(InProcessComponentLauncher):
    def __init__(self, **kwargs):
      super(DummyComponentLauncher, self).__init__(**kwargs)
      ...

    def _run_executor(self, execution_id: int,
                      input_dict: Dict[Text, List[types.Artifact]],
                      output_dict: Dict[Text, List[types.Artifact]],
                      exec_properties: Dict[Text, Any]) -> None:
      ...
  return DummyComponentLauncher
```
which leads to pickling error (https://gist.github.com/iamsusiep/2f61b3277c669734495e3123efe16de7)

Reply:
MyDummyComponentLauncher.set_dummy_executors() shouldn't have any parameters in my thought. Because MyDummyComponentLauncher will be defined in user code, set_dummy_executors still can set required executors from global environment. Although it doesn't look neat...

Reply:
Another idea. 
It uses class variable.

```
import abc
import pickle

class DummyLauncher(abc.ABC):
  def __init__(self):
    self._components = DummyLauncher._components

def create_my_dummy_launcher(components):
  cls = DummyLauncher
  cls._components = components
  return cls
 
my_dummy_launcher = create_my_dummy_launcher(["a","b","c"])

pickle.dump(my_dummy_launcher,open('x.sav','wb'))
my_new_dummy_launcher = pickle.load(open('x.sav','rb'))
print(my_dummy_launcher()._components)
print(my_new_dummy_launcher()._components)
```

Reply:
If we use class variables in the `run_executor()` directly, we won't need `__init__` function in `DummyLauncher`

```
import abc
import pickle

class DummyLauncher(abc.ABC):
  def run_executor(self):
    print(self._components)

def create_my_dummy_launcher(components):
  cls = DummyLauncher
  cls._components = components
  return cls
 
my_dummy_launcher = create_my_dummy_launcher(["a","b","c"])

pickle.dump(my_dummy_launcher,open('x.sav','wb'))
my_new_dummy_launcher = pickle.load(open('x.sav','rb'))
my_dummy_launcher().run_executor()
my_new_dummy_launcher().run_executor()
```